SI-7746 fix unspecifc non-exhaustiveness warnings and non-determinism in pattern matcher (2.11)

gbasler opened the pull request:
This pull request should resolve all remaining non-determinisms regarding exhaustiveness warnings in the pattern matcher. It was buried deep in the machinery and thus not easy to find. The non-determinism is also the reason behind SI-7746.

Review by @adriaanm or @retronym

1) The implementation of the DPLL algorithm was flawed and did not return all solutions. Depending on the ordering of the symbols, more or less solutions were found.
You can find a test project that compares the results of the implementation with Sat4j here:
https://github.com/gbasler/dpll
2) The formulas that were passed to the solver resulted in counter examples that were invalid and removed in a following filtering step. That led to the weird situation where the fixed algorithm returned less counter examples than the buggy one. The recursion goes now deeper than before and also this led to less counter examples reported, since it was stopped too early before.
3) I also cleaned up the `LinkedHashSet` workarounds. They should not be needed anymore. I've repeated Jason's checks (see below) and could not find any problems.

If the search for exhaustivity stops at the max recursion depth, all bets are off since we might not find all possible counter examples. Thus I added a switch `-Ypatmat-exhaust-depth <n> | <off>` that lets the user set the max recursion depth. This is necessary in order to keep the regression tests stable.

I'm working on another PR that will get rid of the "analysis budget" error but this one is a prerequisite to make it work.

Check 1:

```
  ~/Documents/scala-fix-nondet-patmat:ticket/7746$ (f=test/files/neg/t8430.scala; (for i in {1..20}; do echo $f; done; printf "\n") | qbin/scalac -Xresident) 2>&1 | grep "fail" | sort | uniq | wc -l
      1
```

Check 2:

```
 for i in {1..100}; do cp test/files/neg/t7020.scala test/files/neg/t7020-$i.scala; cp test/files/neg/t7020.flags test/files/neg/t7020-$i.flags; cat test/files/neg/t7020.check | sed "s/t7020/t7020-$i/" > test/files/neg/t7020-$i.check; done

~/Documents/scala-fix-nondet-patmat:ticket/7746$ test/partest --terse --srcpath disabled test/files/neg/t7020*.scala

Selected 101 tests drawn from specified tests


# starting 101 tests in neg

........................................................................

.............................


101/101 passed (elapsed time: 00:00:32)

Test Run PASSED
```

gbasler committed:
Cleanup `LinkedHashSet` fixes and replace them with `Set` (i.e., back
to initial implementation).

Assuming that the DPLL procedure does not run into max recursion
depth, that workaround is not needed anymore, since the non-
determinism has been fixed.

gbasler commented:
PLS REBUILD/pr-scala@c5b1cf2

scala-jenkins commented:
(kitty-note-to-self: ignore 54028901)
:cat: Roger! Rebuilding pr-scala for c5b1cf2f. :rotating_light:

gbasler commented:
PLS REBUILD/pr-scala@c5b1cf2

scala-jenkins commented:
(kitty-note-to-self: ignore 54038510)
:cat: Roger! Rebuilding pr-scala for c5b1cf2f. :rotating_light:

gbasler commented:
@retronym the build is failing in scala-ide. Could it be the same issue as in #3948 ? (spurious error)

gbasler commented:
PLS REBUILD/pr-scala@0d87ef6

scala-jenkins commented:
(kitty-note-to-self: ignore 54094540)
:cat: Roger! Rebuilding pr-scala for 0d87ef6f. :rotating_light:

gbasler commented:
@retronym could you review it, since Adriaan seems to be away?

retronym commented:
Could you please tighten up the word-wrapping in the commit messages.

The first should be:

```
SI-7746 patmat: fix non-determinism, infeasible counter examples

Fixes non-determinism within the DPLL algorithm and disallows
infeasible counter examples directly in the formula.

The function to compute all solutions was flawed and thus only
returned a subset of the solutions. The algorithm would stop too soon
and thus depending on the ordering of the symbols return more or less
solutions. I also added printing a warning when the search was stopped
because the max recursion depth was reached. This is very useful as an
explanation of spuriously failing regression tests, since less counter
examples might be reported. In such a case the recursion depth should
be set to infinite by adding `-Ypatmat-exhaust-depth off`.

The formulas that are passed to the solver are now restricted in order
to avoid generating invalid counter examples. Previously the
incorrectly set up formula resulted in invalid counter examples that
were then removed in the following filtering step. That led to the
weird situation where the fixed DPLL algorithm returned less counter
examples than the buggy version.
```

retronym commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -228,6 +228,8 @@ trait ScalaSettings extends AbsScalaSettings
   val Ytyperdebug             = BooleanSetting("-Ytyper-debug", "Trace all type assignments.")
   val Ypatmatdebug            = BooleanSetting("-Ypatmat-debug", "Trace pattern matching translation.")
+  val YpatmatDPLLdepth        = IntSetting("-Ypatmat-exhaust-depth", "off", 20, Some((10, Int.MaxValue)),
+    str => Some(if(str == "off") Int.MaxValue else str.toInt))
```
`str.equalsIgnoreCase("off")`

retronym replied:
The variable name (`YpatmatDPLLdepth`) is different from the option name (`-Ypatmat-exhaust-depth`). Intentional?

gbasler replied:
I forgot to rename the variable as well.

retronym commented on src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:
```diff
@@ -204,7 +207,7 @@ trait Logic extends Debugging  {
       val start = if (Statistics.canEnable) Statistics.startTimer(patmatAnaVarEq) else null
 
-      val vars = mutable.LinkedHashSet[Var]()
+      val vars = new scala.collection.mutable.HashSet[Var]
```
`new mutable.HashSet[Var]` (keep using the import)

retronym commented on src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:
```diff
@@ -486,7 +498,15 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
+        _.toSeq.combinations(2).toIndexedSeq.filterNot { case Seq(a, b) =>
+            val aIsSubTypeOfb = a.const.tp <:< b.const.tp
+            val bIsSubTypeOfa = b.const.tp <:< a.const.tp
+            aIsSubTypeOfb || bIsSubTypeOfa
```
``` scala
def isSub(sym1: Sym, sym2: Sym) = sym1.const.tp <:< sym2.const.tp
isSub(a, b) || isSub(b, a)
```

retronym replied:
Can you think of a pattern match (perhaps one with many cases, or one with symbols that have large domains) that might lead to this part being a performance bottleneck?

gbasler replied:
Yes that's easy to construct: just create a symbol with a large domain. So the number of subtype checks is quadratic in the number of symbols in the domain. 

I added now a fast path in case a symbol has no children and avoid the subtype check in that case.

I needed the subtype check to be able to report one (redundant) case for `patmatexhaust.scala`. See also the comments in the code. Maybe it can just be omitted (or omitted if the domain is larger than a specific number)? 

In the end I need this list to add more constraints to the CNF formula that is checked. So the number of clauses is quadratic in the number of symbols as well. I don't think this can be avoided.
However, such a test case would - with the current implementation - most likely run into a `CNF budget exceeded` error anyway.

retronym replied:
How does `test/files/pos/t8531` perform before and after this change?

gbasler replied:
It terminated rather quickly in both cases, so I did:

```
cat Test.scala > Test-Big.scala;for i in {1..100}; do echo "class Foo$i {" >> Test-Big.scala ; cat Test.scala |  tail -n 17 >> Test-Big.scala ; done
```

Master:

```
~/Documents/scala-2.12-master/sandbox:2.11.x$ time ../qbin/scalac -Xfatal-warnings -unchecked Test-Big.scala MyEnum.java 

real    0m9.913s
user    0m22.006s
sys 0m0.458s
```

Branch:

```
~/Documents/scala-fix-nondet-patmat/sandbox:ticket/7746-2.11$ time ../qbin/scalac -Xfatal-warnings -unchecked Test-Big.scala MyEnum.java 

real    0m10.053s
user    0m22.483s
sys 0m0.443s
```

Memory usage was about 170MB in both cases.

retronym commented on test/files/neg/t7020.check:
```diff
@@ -1,17 +1,17 @@
 t7020.scala:3: warning: match may not be exhaustive.
-It would fail on the following inputs: List((x: Int forSome x not in (1, 2, 4, 5, 6, 7))), List((x: Int forSome x not in (1, 2, 4, 5, 6, 7)), _), List(1, _), List(2, _), List(4, _), List(5, _), List(6, _), List(7, _), List(_, _)
+It would fail on the following inputs: List((x: Int forSome x not in (1, 2, 4, 5, 6, 7))), List((x: Int forSome x not in (1, 2, 4, 5, 6, 7)), _), List(1, _), List(2, _), List(4, _), List(5, _), List(6, _), List(7, _), List(??, _)
```
Do you know what `??` means in this output, and why this changed?

gbasler replied:
If I add `println(s"== ${equalTo}, != ${notEqualTo} => $res")` to `MatchAnalysis` I see the output below.  I'll investigate a bit more, why `List(??, _)` occurs. What counter examples would you expect? Did the original output make sense?

```
~/Documents/scala-fix-nondet-patmat/sandbox:ticket/7746-2.11$ ../qbin/scalac -Xfatal-warnings -unchecked t7020.scala 
== List(), != List(2, 7, 6, 1, 5, 4) => (x: Int forSome x not in (1, 2, 4, 5, 6, 7))
== List(Nil), != List(scala.collection.immutable.::[?]) => Nil
== List(scala.collection.immutable.::[?]), != List(Nil) => List((x: Int forSome x not in (1, 2, 4, 5, 6, 7)))
== List(2), != List(7, 6, 1, 5, 4) => 2
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(2, _)
== List(2, 7), != List(6, 1, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7), != List(2, 6, 1, 5, 4) => 7
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(7, _)
== List(2, 7, 5), != List(6, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 5), != List(7, 6, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 5), != List(2, 6, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(5), != List(2, 7, 6, 1, 4) => 5
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(5, _)
== List(2, 7, 1, 5), != List(6, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 1), != List(6, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 1, 5), != List(7, 6, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 1), != List(7, 6, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 1, 5), != List(2, 6, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 1), != List(2, 6, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(1, 5), != List(2, 7, 6, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(1), != List(2, 7, 6, 5, 4) => 1
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(1, _)
== List(2, 7, 6, 1, 5), != List(4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6, 1), != List(5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6, 5), != List(1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6), != List(1, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 1, 5), != List(7, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 1), != List(7, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 5), != List(7, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6), != List(7, 1, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 1, 5), != List(2, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 1), != List(2, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 5), != List(2, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6), != List(2, 1, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 1, 5), != List(2, 7, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 1), != List(2, 7, 5, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 5), != List(2, 7, 1, 4) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6), != List(2, 7, 1, 5, 4) => 6
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(6, _)
== List(2, 7, 6, 1, 5, 4), != List() => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6, 1, 4), != List(5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6, 5, 4), != List(1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 6, 4), != List(1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 1, 5, 4), != List(6) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 1, 4), != List(6, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 5, 4), != List(6, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 7, 4), != List(6, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 1, 5, 4), != List(7) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 1, 4), != List(7, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 5, 4), != List(7, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 6, 4), != List(7, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 1, 5, 4), != List(7, 6) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 1, 4), != List(7, 6, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 5, 4), != List(7, 6, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(2, 4), != List(7, 6, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 1, 5, 4), != List(2) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 1, 4), != List(2, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 5, 4), != List(2, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 6, 4), != List(2, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 1, 5, 4), != List(2, 6) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 1, 4), != List(2, 6, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 5, 4), != List(2, 6, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(7, 4), != List(2, 6, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 1, 5, 4), != List(2, 7) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 1, 4), != List(2, 7, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 5, 4), != List(2, 7, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(6, 4), != List(2, 7, 1, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(1, 5, 4), != List(2, 7, 6) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(1, 4), != List(2, 7, 6, 5) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(5, 4), != List(2, 7, 6, 1) => ??
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(??, _)
== List(4), != List(2, 7, 6, 1, 5) => 4
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List(4, _)
== List(), != List(2, 7, 6, 1, 5, 4) => (x: Int forSome x not in (1, 2, 4, 5, 6, 7))
== List(scala.collection.immutable.::[?]), != List(Nil) => List(_)
== List(scala.collection.immutable.::[?]), != List(Nil) => List((x: Int forSome x not in (1, 2, 4, 5, 6, 7)), _)
```

retronym commented on src/compiler/scala/tools/nsc/transform/patmat/Solving.scala:
```diff
@@ -141,36 +134,72 @@ trait Solving extends Logic {
+          global.reporter.warning(scala.reflect.internal.util.NoPosition,
+            "(Unreachability analysis reached max recursion depth, not all missing cases are reported. " +
+              s"Please try with scalac -Ypatmat-exhaust-depth ${maxDPLLdepth * 2} or -Ypatmat-exhaust-depth off.)")
+          models
```
We'll need to find a way to report this at a suitable position. The caller, `MatchAnalysis`, uses `prevBinder.pos` for converting thrown `AnalysisBudget.Exception`s into positioned warnings. But I see you can't throw an exception here, as you actually continue processing.

Maybe you can add a new method `def reportWarning(message: String)` to the interface `PropositionalLogic`, and implement that in `OptimizingMatchTranslator` using `typer.context.tree.position` as the position.

gbasler replied:
Your suggestion is great, I was struggling with this!

gbasler committed:
SI-7746 patmat: fix non-determinism, infeasible counter examples
Fixes non-determinism within the DPLL algorithm and disallows
infeasible counter examples directly in the formula.

The function to compute all solutions was flawed and thus only
returned a subset of the solutions. The algorithm would stop too soon
and thus depending on the ordering of the symbols return more or less
solutions. I also added printing a warning when the search was stopped
because the max recursion depth was reached. This is very useful as an
explanation of spuriously failing regression tests, since less counter
examples might be reported. In such a case the recursion depth should
be set to infinite by adding `-Ypatmat-exhaust-depth off`.

The mapping of the solutions of the DPLL algorithm to counter examples
has been adapted to take the additional solutions from the
solver into account:

Consider for example `t8430.scala`:

```Scala
sealed trait CL3Literal
case object IntLit extends CL3Literal
case object CharLit extends CL3Literal
case object BooleanLit extends CL3Literal
case object UnitLit extends CL3Literal

sealed trait Tree
case class LetL(value: CL3Literal) extends Tree
case object LetP extends Tree
case object LetC extends Tree
case object LetF extends Tree

object Test {
  (tree: Tree) => tree match {case LetL(CharLit) => ??? }
}
```

This test contains 2 domains,  `IntLit, CharLit, ...` and `LetL, LetP, ...`,
the corresponding formula to check exhaustivity looks like:

```
    V1=LetC.type#13    \/ V1=LetF.type#14 \/    V1=LetL#11     \/  V1=LetP.type#15    /\
 V2=BooleanLit.type#16 \/  V2=CharLit#12  \/ V2=IntLit.type#17 \/ V2=UnitLit.type#18  /\
      -V1=LetL#11      \/ -V2=CharLit#12  \/                   \/
```

The first two lines assign a value of the domain to the scrutinee (and
the correponding member in case of `LetL`) and prohibit the counter
example `LetL(CharLit)` since it's covered by the pattern match. The
used Boolean encoding allows that scrutinee `V1` can be equal to
`LetC` and `LetF` at the same time and thus, during enumeration of all
possible solutions of the formula, such a solution will be found,
since only one literal needs to be set to true, to  satisfy that
clause. That means, if at least one of the literals of such a clause
was in the `unassigned` list of the DPLL procedure, we will get
solutions where the scrutinee is equal to more than one element of the
domain.

A remedy would be to add constraints that forbid both literals
to be true at the same time. His is infeasible for big domains (see
`pos/t8531.scala`), since we would have to add a quadratic number of
clauses (one clause for each pair in the domain). A much simpler
solution is to just filter the invalid results. Since both values for
`unassigned` literals are explored, we will eventually find a valid
counter example.

gbasler commented:
@retronym Thanks for the thorough review! You pointed out some weak points. I've found a more elegant solution to deal with the solutions of the solver and removed the additional constrains. So the size of the formulas is not increased anymore. I could also fix the result of `t7020`.

retronym commented:
As discussed offline, I'd love if you could add a commit with some more internal documentation to capture what you've learned about the existing code and to explain the changes (especially in `expandUnassigned`).

I'd like to leave this PR open until @adriaanm is back on deck so he can complete the review.

gbasler commented:
PLS REBUILD/pr-scala@663dfd8

scala-jenkins commented:
(kitty-note-to-self: ignore 55859867)
:cat: Roger! Rebuilding pr-scala for 663dfd81. :rotating_light:

gkossakowski commented:
@adriaanm touched the ground in the old country so I'm assigning to him. Expect some delays in review due to :coffee: around there never coming up to @adriaanm's standards, though. :bowtie:

adriaanm commented:
This is awesome -- thank you! Better than an espresso made by a mustached hipster :-)
Starting my review now.

adriaanm commented:
Oops, looks like I reviewed an [outdated commit](https://github.com/gbasler/scala/commit/663dfd81e07538934ab1c12fd687d8d0e8ca1178).

gbasler commented:
@adriaanm I hope you found some decent :coffee: - with or without being served by a moustached hipster ;-)

I force pushed yesterday evening and addressed your comments. You reviewed the correct commit. Maybe I should just have added a new commit and squashed later...

adriaanm commented:
I decided to take matters into my own hands. On my way home with some beans. Will review by the end of the (Belgian) day.

adriaanm commented:
LGTM -- my remaining follow-up questions are academic

Thanks, @gbasler, for fixing a long standing issue that completely eluded me!

adriaanm commented:
Thanks: https://twitter.com/adriaanm/status/519144318079356928 :)

som-snytt commented:
"buried deep in the machinery" makes a great tweet.

I still haven't found the "share on twitter" button, but maybe they'll make it possible to comment on PRs via twitter.
